core/extensions/impl/storage_extension.cpp

```diff
@@ -3,99 +3,188 @@
-    std::terminate();
+  void StorageExtension::ext_set_storage(const runtime::WasmPointer key_data,
+                                         runtime::SizeType key_length,
+                                         const runtime::WasmPointer value_data,
+                                         runtime::SizeType value_length) {
+    auto key = memory_->loadN(key_data, key_length);
+    auto value = memory_->loadN(value_data, value_length);
+    auto put_result = db_->put(key, value);
+    if (not put_result) {
+      logger_->warn(
+          "ext_set_storage failed, due to fail in trie db with reason: {}",
+          put_result.error().message());
+    }
   }
 
-  uint8_t *StorageExtension::ext_get_allocated_storage(const uint8_t *key_data,
-                                                       uint32_t key_length,
-                                                       uint32_t *written) {
-    std::terminate();
+  // -------------------------Trie operations--------------------------
+
+  void StorageExtension::ext_blake2_256_enumerated_trie_root(
+      runtime::WasmPointer values_data, runtime::WasmPointer lens_data,
+      runtime::SizeType lens_length, runtime::WasmPointer result) {
+    std::vector<common::Buffer> data;
+    for (runtime::SizeType i = 0; i < lens_length; i++) {
+      auto length = memory_->load32s(lens_data + i * 4);
+      data.emplace_back(memory_->loadN(values_data, length));
+      values_data += length;
+    }
+
+    auto root = codec_->orderedTrieRoot(data);
+    memory_->storeBuffer(
+        result, common::Buffer(std::vector<uint8_t>(root.begin(), root.end())));
   }
 
-  uint32_t StorageExtension::ext_get_child_storage_into(
-      const uint8_t *storage_key_data, uint32_t storage_key_length,
-      const uint8_t *key_data, uint32_t key_length, uint8_t *value_data,
-      uint32_t value_length, uint32_t value_offset) {
-    std::terminate();
+  runtime::SizeType StorageExtension::ext_storage_changes_root(
+      runtime::WasmPointer parent_hash_data, runtime::SizeType parent_hash_len,
+      runtime::SizeType parent_num, runtime::WasmPointer result) {
+    // TODO Implement PRE-95, kamilsa, 03.04.2019
+    logger_->error("Unimplemented");
+    return 0;
   }
 
-  uint32_t StorageExtension::ext_get_storage_into(const uint8_t *key_data,
-                                                  uint32_t key_length,
-                                                  uint8_t *value_data,
-                                                  uint32_t value_length,
-                                                  uint32_t value_offset) {
+  void StorageExtension::ext_storage_root(runtime::WasmPointer result) const {
+    const auto root = db_->getRoot();
+    memory_->storeBuffer(result, root);
+  }
+
+  std::optional<common::Buffer> StorageExtension::get(
+      const common::Buffer &key, runtime::SizeType offset,
+      runtime::SizeType max_length) const {
+    const auto data = db_->get(key);
```

Comment:
change `get` to `outcome::result` and use OUTCOME_TRY here.

also change std::optional -> outcome::result